### PR TITLE
Improve testing of MySQL functions.

### PR DESCRIPTION
**Pull request by maxolasersquad**

### CODE REVIEW

**Comment by maxolasersquad** [comment]
Still working on this. Creating the PR early to make sure the tests pass via Travis on all supported setups and to get any feedback early.

**Comment by stevelacey** [comment]
I support any improvements to the test suite, the coverage here is bad.

Only feedback I have so far is that your changes miss the `DoctrineExtensions\Tests` prefix to the namespaces.

**Comment by maxolasersquad** [comment]
@stevelacy, would you look at the failing test and recommend a solution? It seems that Doctrine 2.1 and 2.2 are generating SQL than Doctrine 2.3+. I am thinking that maybe this represents a bug in how the Binary extension is constructing the SQL. Thoughts?

**Comment by maxolasersquad** [comment]
I have completed writing tests for all of the MySQL extensions. It has exposed bugs in at least two of them for older versions of Doctrine. I'm going to attempt to fix the problems.

**Comment by maxolasersquad** [comment]
There is different behavior in Doctrine <= 2.2 as it deals with parsed strings. I have put in some code that checks the result of Parser::StringPrimary() and provides backwards compatibility. All the MySQL functions have tests, and all of the tests pass.

**Comment by maxolasersquad** [comment]
@stevelacey, if you find some time would give this a review?

**Comment by stevelacey** [comment]
@maxolasersquad ah I suck, sorry, I am looking now

**Comment by stevelacey** [comment]
@maxolasersquad I am just giving this some through with regards to the 2.1/2.2 fixes – I am more than happy to merge all of tests, this is great, but wondering whether it's even worth supporting these older versions of Doctrine anymore

**Comment by stevelacey** [comment]
@maxolasersquad if we do keep these fixes...

I would like to see this fix either explicitly check the Doctrine version and alter behaviour based on that or reverse it to test for whether the value is a string rather than for Doctrine's `Node` class, meaning this is less likely to break in future versions and more explicitly addressing the old Doctrine problem.

It might also be better to move this fix to an abstract class for reuse – perhaps extend `FunctionNode` with a `(get|coerse)StringPrimary($value)` method or something if this is the only place it's needed.

**Comment by maxolasersquad** [comment]
I took the route of explicitly checking the version of Doctrine because it is self documenting (I should have put comments in my previous code) and because if Doctrine <= 2.2 is removed from support it makes finding and removing this code simple. I'll leave what versions of Doctrine and PHP it makes sense to support up to the people supporting this library long-term.

**Comment by stevelacey** [comment]
I would still prefer this to be abstracted into a a base class, makes removal even easier + DRYer :)

File: tests/Query/Mysql/DateDiffTest.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+            $q->getSql()
+        );
+    }
+
+    public function testDateDiffWithColumnAlias()
+    {
+        $q = $this->entityManager->createQuery("SELECT DATEADD(p.created, 1, 'DAY') as modified_date FROM DoctrineExtensions\Tests\Entities\Date as p WHERE DATEDIFF(CURRENT_TIME(), modified_date) < 7");
+
+        $this->assertEquals(
+            "SELECT DATE_ADD(d0_.created, INTERVAL 1 DAY) AS sclr0 FROM Date d0_ WHERE DATEDIFF(CURRENT_TIME, sclr0) < 7",
```
Comment (eigan) [inline tests/Query/Mysql/DateDiffTest.php]: Sorry, my test was wrong. This should be HAVING, not WHERE. alias in WHERE is not valid SQL.

http://stackoverflow.com/questions/9080128/how-do-i-reference-an-alias-in-a-where-clause

**Comment by stevelacey** [comment]
@maxolasersquad any progress here? would be good to get this merged

**Comment by maxolasersquad** [comment]
I'm not sure why, but I thought this had already been merged. I'll see where I can find some time to work on this. It will probably be a couple days before I can address it.

**Comment by maxolasersquad** [comment]
All tests are passing now.

File: tests/Query/Mysql/DateDiffTest.php [hunk 2]
```diff
@@ -18,9 +19,14 @@ public function testDateDiff()
     public function testDateDiffWithColumnAlias()
     {
         $q = $this->entityManager->createQuery("SELECT DATEADD(p.created, 1, 'DAY') as modified_date FROM DoctrineExtensions\Tests\Entities\Date as p WHERE DATEDIFF(CURRENT_TIME(), modified_date) < 7");
+        if (Version::VERSION >= 2.5) {
+            $alias = 'sclr_0';
+        } else {
+            $alias = 'sclr0';
+        }
```
Comment (stevelacey) [inline tests/Query/Mysql/DateDiffTest.php]: We actually handle this globally with a custom `assertEquals` – basically use the new syntax (with `_`) and it'll get stripped in the tests on older versions.

https://github.com/beberlei/DoctrineExtensions/blob/master/tests/Query/DbTestCase.php#L31-L39

Comment (maxolasersquad) [inline tests/Query/Mysql/DateDiffTest.php]: That makes sense. I was wondering why I had to do that for this one test.

**Comment by stevelacey** [comment]
@maxolasersquad would still like to merge this – is the other stuff that alternates tests around the Doctrine version necessary?

**Comment by stevelacey** [comment]
@maxolasersquad finally merged this, thanks for your contribution 😀

**Pull request closed without merging**